### PR TITLE
YJIT: Support splat with C methods with -1 arity

### PR DESCRIPTION
**Pull request by XrXr**
Usually we deal with splats by speculating that they're of a specific
size. In this case, the C method takes a pointer and a length, so
we can support changing sizes just fine.

On lobsters:

```
num_send_dynamic:            727,218 ( 2.6%)
num_send_dynamic:            684,529 ( 2.5%)
```

### CODE REVIEW

**Review by maximecb — APPROVED** [review]
Open to merging it, but my question would be, is this faster than the dynamic fallback since it's adding two C calls?

**Comment by XrXr** [comment]
> is this faster than the dynamic fallback?

```
Calculating -------------------------------------
                     yjit-pre --yjit  yjit-post --yjit 
       var_len_splat         33.948M           35.412M i/s -     10.000M times in 0.294571s 0.282392s

Comparison:
                    var_len_splat
    yjit-post --yjit:  35411768.0 i/s 
     yjit-pre --yjit:  33947673.1 i/s - 1.04x  slower

$ cat ../../splat_neg.yml 
prelude:
  splat = [0, 1]
benchmark:
  var_len_splat: |
    splat.fetch(*splat)
loop_count: 10_000_000
```

Yes, even with the C calls we still win out by having static calls and no overhead for falling back.

File: yjit.c [hunk 1]
```diff
@@ -890,6 +890,51 @@ rb_yjit_ruby2_keywords_splat_p(VALUE obj)
     return FL_TEST_RAW(last, RHASH_PASS_AS_KEYWORDS);
 }
 
+// Checks to establish preconditions for rb_yjit_splat_varg_cfunc()
+VALUE
+rb_yjit_splat_varg_checks(VALUE *sp, VALUE splat_array, rb_control_frame_t *cfp)
+{
+    if (!RB_TYPE_P(splat_array, T_ARRAY)) return Qfalse;
```
Comment (k0kubun) [inline yjit.c:897]: Is this something we could optimize away if we do this using `guard_object_is_array`, or is the information often wiped out on benchmarks?

**Review by k0kubun — APPROVED** [review]

File: yjit.c (hunk 1 above)
Comment (XrXr) [reply yjit.c:897]: It's optimizable, yeah. I'll toss it in.

**Pull request merged**